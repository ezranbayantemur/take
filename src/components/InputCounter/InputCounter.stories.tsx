import {InputCounter} from '@components';
import type {Meta, StoryObj} from '@storybook/react-native';
import type {InputCounterProps} from './InputCounter.types';

const InputMeta: Meta<InputCounterProps> = {
  argTypes: {
    onValueChange: {
      action: 'on value change',
    },
  },
  args: {
    value: 1,
  },
  title: 'InputCounter',
  component: InputCounter,
};

export default InputMeta;

export const Basic: StoryObj<InputCounterProps> = {};
