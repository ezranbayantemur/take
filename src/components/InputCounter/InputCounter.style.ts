import {StyleSheet} from 'react-native';
import {color, padding, font, typography} from '@style';

export default StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
  button_container: {
    width: 25,
    height: 25,
    backgroundColor: color.MONZA,
    marginHorizontal: padding.SMALL,
    borderRadius: 20,
    justifyContent: 'center',
    alignItems: 'center',
  },
  button_text: {
    color: color.ALABASTER,
    fontFamily: font.BLACK,
    fontSize: typography.NORMAL,
  },
  value: {
    fontFamily: font.BLACK,
    fontSize: typography.MEDIUM,
  },
});
