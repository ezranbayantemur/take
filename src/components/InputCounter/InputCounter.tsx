import React from 'react';
import {Text, TouchableOpacity, View} from 'react-native';
import styles from './InputCounter.style';
import type {InputCounterProps} from './InputCounter.types';

const Input = ({
  testID,
  value,
  onIncrease,
  onDecrease,
  onValueChange,
}: InputCounterProps) => {
  const [innerValue, setInnerValue] = React.useState(value);

  React.useEffect(() => {
    if (onValueChange) {
      onValueChange(innerValue);
    }
  }, [innerValue, onValueChange]);

  const handleIncrease = () => {
    const newValue = innerValue + 1;
    if (onIncrease) {
      onIncrease(newValue);
    }
    setInnerValue(newValue);
  };

  const handleDecrease = () => {
    const newValue = innerValue - 1;
    if (onDecrease) {
      onDecrease(newValue);
    }
    setInnerValue(newValue);
  };

  return (
    <View style={styles.container}>
      <TouchableOpacity
        testID={`${testID}_increase_touchable`}
        style={styles.button_container}
        onPress={handleIncrease}>
        <Text style={styles.button_text}>+</Text>
      </TouchableOpacity>
      <Text testID={`${testID}_value_text`} style={styles.value}>
        {innerValue}
      </Text>
      <TouchableOpacity
        testID={`${testID}_decrease_touchable`}
        style={styles.button_container}
        onPress={handleDecrease}>
        <Text style={styles.button_text}>-</Text>
      </TouchableOpacity>
    </View>
  );
};

export default React.memo(Input);
