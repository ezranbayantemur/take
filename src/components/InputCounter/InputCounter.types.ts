export interface InputCounterProps {
  testID: string;
  value: number;
  onIncrease?: (value: number) => void;
  onDecrease?: (value: number) => void;
  onValueChange?: (value: number) => void;
}
