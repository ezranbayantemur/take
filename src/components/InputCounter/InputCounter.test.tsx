import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import InputCounter from './InputCounter';
jest.useFakeTimers();

const mockOnDecrease = jest.fn();
const mockOnIncrease = jest.fn();
const mockOnValueChange = jest.fn();

let wrapper: ReturnType<typeof render>;
describe('Input unit tests', () => {
  beforeEach(() => {
    wrapper = render(
      <InputCounter
        testID="test"
        value={5}
        onDecrease={mockOnDecrease}
        onIncrease={mockOnIncrease}
        onValueChange={mockOnValueChange}
      />,
    );

    mockOnDecrease.mockClear();
    mockOnIncrease.mockClear();
    mockOnValueChange.mockClear();
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render value correctly', () => {
    expect(wrapper.queryByText('5')).not.toBeNull();
  });

  it('should increase value correctly', () => {
    const touchable = wrapper.getByTestId('test_increase_touchable');
    fireEvent(touchable, 'onPress');

    expect(wrapper.queryByText('6')).not.toBeNull();
  });

  it('should decrease value correctly', () => {
    const touchable = wrapper.getByTestId('test_decrease_touchable');
    fireEvent(touchable, 'onPress');

    expect(wrapper.queryByText('4')).not.toBeNull();
  });

  it('should trigger onIncrease correctly', () => {
    const touchable = wrapper.getByTestId('test_increase_touchable');
    fireEvent(touchable, 'onPress');

    expect(mockOnIncrease).toBeCalledWith(6);
  });

  it('should trigger onDecrease correctly', () => {
    const touchable = wrapper.getByTestId('test_decrease_touchable');
    fireEvent(touchable, 'onPress');

    expect(mockOnDecrease).toBeCalledWith(4);
  });

  it('should trigger onValueChange correctly', () => {
    fireEvent(wrapper.getByTestId('test_increase_touchable'), 'onPress');
    expect(mockOnValueChange).toBeCalledWith(6);

    fireEvent(wrapper.getByTestId('test_decrease_touchable'), 'onPress');
    expect(mockOnValueChange).toBeCalledWith(5);
  });

  it('should increase without onIncrease prop correctly', () => {
    wrapper.rerender(<InputCounter testID="test" value={5} />);
    const touchable = wrapper.getByTestId('test_increase_touchable');
    fireEvent(touchable, 'onPress');

    expect(wrapper.queryByText('6')).not.toBeNull();
  });

  it('should decrease without onDecrease correctly', () => {
    wrapper.rerender(<InputCounter testID="test" value={5} />);
    const touchable = wrapper.getByTestId('test_decrease_touchable');
    fireEvent(touchable, 'onPress');

    expect(wrapper.queryByText('4')).not.toBeNull();
  });
});
