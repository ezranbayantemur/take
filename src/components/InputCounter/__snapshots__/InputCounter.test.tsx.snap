// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input unit tests should render correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#DB040E",
        "borderRadius": 20,
        "height": 25,
        "justifyContent": "center",
        "marginHorizontal": 8,
        "opacity": 1,
        "width": 25,
      }
    }
    testID="test_increase_touchable"
  >
    <Text
      style={
        {
          "color": "#FBFBFB",
          "fontFamily": "Lato-Black",
          "fontSize": 15,
        }
      }
    >
      +
    </Text>
  </View>
  <Text
    style={
      {
        "fontFamily": "Lato-Black",
        "fontSize": 17,
      }
    }
    testID="test_value_text"
  >
    5
  </Text>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#DB040E",
        "borderRadius": 20,
        "height": 25,
        "justifyContent": "center",
        "marginHorizontal": 8,
        "opacity": 1,
        "width": 25,
      }
    }
    testID="test_decrease_touchable"
  >
    <Text
      style={
        {
          "color": "#FBFBFB",
          "fontFamily": "Lato-Black",
          "fontSize": 15,
        }
      }
    >
      -
    </Text>
  </View>
</View>
`;
